applications/nrf5340_audio/src/modules/music_led_sync.c

```diff
@@ -0,0 +1,161 @@
+		 * as every other element in the array is the belonging imaginary part.
+		 */
+		if (fft_data[(i + 1) * 2] > 0) {
+			/* Upscaling of values according to CMSIS doc */
+			if (i < CONFIG_RGB_RED_FREQ_BIN_RANGE) {
+				frequency_modules[0] += (fft_data[i] << 7);
+			} else if (i < CONFIG_RGB_GREEN_FREQ_BIN_RANGE) {
+				frequency_modules[1] += (fft_data[i] << 7);
+			} else {
+				frequency_modules[2] += (fft_data[i] << 7);
+			}
+		}
+	}
+
+	/* Reducing intensity of less dominant modules*/
+	if ((frequency_modules[0] > frequency_modules[1]) &&
+	    (frequency_modules[0] > frequency_modules[2])) {
+		frequency_modules[1] *= 0.5;
+		frequency_modules[2] *= 0.5;
+	} else if ((frequency_modules[1] > frequency_modules[0]) &&
+		   (frequency_modules[1] > frequency_modules[2])) {
+		frequency_modules[0] *= 0.5;
+		frequency_modules[2] *= 0.5;
+	} else if ((frequency_modules[2] > frequency_modules[0]) &&
+		   frequency_modules[2] > frequency_modules[1]) {
+		frequency_modules[0] *= 0.5;
+		frequency_modules[1] *= 0.5;
+	}
+
+	/* Calculating duty cycle values, +1 added to prevent division by 0 */
+	uint32_t total_modules_val =
+		frequency_modules[0] + frequency_modules[1] + frequency_modules[2] + 1;
+
+	red_pwm_led_duty_cycle =
+		(rgb_led_period - (rgb_led_period / total_modules_val * frequency_modules[0]));
+	green_pwm_led_duty_cycle =
+		(rgb_led_period - (rgb_led_period / total_modules_val * frequency_modules[1]));
+	blue_pwm_led_duty_cycle =
+		(rgb_led_period - (rgb_led_period / total_modules_val * frequency_modules[2]));
+
+	ret = pwm_set_pulse_dt(&red_pwm_led, red_pwm_led_duty_cycle);
+	if (ret) {
+		LOG_DBG("Failed to write to red RGB, ret: %d", ret);
+	}
+
+	ret = pwm_set_pulse_dt(&green_pwm_led, green_pwm_led_duty_cycle);
+	if (ret) {
+		LOG_DBG("Failed to write to green RGB, ret: %d", ret);
+	}
+
+	ret = pwm_set_pulse_dt(&blue_pwm_led, blue_pwm_led_duty_cycle);
+	if (ret) {
+		LOG_DBG("Failed to write to blue RGB, ret: %d", ret);
+	}
+}
+
+static int music_led_sync_thread(void *dummy1, void *dummy2, void *dummy3)
+{
+	arm_rfft_instance_q15 fft_instance;
+	arm_status status;
+	q15_t input[CONFIG_FFT_SAMPLE_SIZE];
+	/* Output has to be twice the size of input */
+	q15_t output[CONFIG_FFT_SAMPLE_SIZE * 2];
+
+	status = arm_rfft_init_q15(&fft_instance, CONFIG_FFT_SAMPLE_SIZE,
+				   CONFIG_FFT_FLAG_INVERSE_TRANSFORM, CONFIG_FFT_FLAG_BIT_REVERSAL);
+	if (status != ARM_MATH_SUCCESS) {
+		LOG_ERR("Failed to init FFT instance - (err: %d)", status);
+		return status;
+	}
+
+	while (1) {
+		k_sem_take(&music_led_sync_update_sem, K_FOREVER);
+		k_sem_take(&music_led_sync_value_sem, K_FOREVER);
+		memcpy(input, pcm_data_mono, CONFIG_FFT_SAMPLE_SIZE);
+		k_sem_give(&music_led_sync_value_sem);
+
+		arm_rfft_q15(&fft_instance, (q15_t *)pcm_data_mono, output);
+		arm_abs_q15(output, output, (CONFIG_FFT_SAMPLE_SIZE * 2));
+		music_led_sync_fft_analyse(output, CONFIG_FFT_SAMPLE_SIZE);
+		k_msleep(125);
```

Comment:
Why is there an k_msleep here? Forgive me if I commented on this one before.

Reply:
This is added to give smoother transition of color for the RGB LED.

Reply:
OK. Then I suggest you create a timer https://docs.zephyrproject.org/3.0.0/reference/kernel/timing/timers.html
and in the `my_expiry_function(struct k_timer *timer_id);` you can add a `k_sem_give` which releases the `while` loop